### PR TITLE
feat: refactor the node key as `version + path`

### PR DESCRIPTION
**Pull request by cool-develope**
ref: #571, #608 

## Context

As discussed in #608, it suggests the node key format as `version + path` instead of `hash`.  It introduces a new way to get new nodes and orphans and remove the root and orphans from storage.

## What does this PR do?

- Refactor the node key with `version + path`
- Remove the root store
- Include #646

### CODE REVIEW

**Comment by tac0turtle** [comment]
@cool-develope can you give a short summary of all the changes in the pr in the description

File: benchmarks/bench_test.go [hunk 1]
```diff
@@ -308,7 +308,7 @@ func BenchmarkLevelDBLargeData(b *testing.B) {
 		{"goleveldb", 50000, 100, 32, 100},
 		{"goleveldb", 50000, 100, 32, 1000},
 		{"goleveldb", 50000, 100, 32, 10000},
-		{"goleveldb", 50000, 100, 32, 100000},
+		// {"goleveldb", 50000, 100, 32, 100000},
```
Comment (tac0turtle) [inline benchmarks/bench_test.go:311]:
```suggestion
		 {"goleveldb", 50000, 100, 32, 100000},
```
?

File: immutable_tree.go [hunk 2]
```diff
@@ -149,8 +149,7 @@ func (t *ImmutableTree) Has(key []byte) (bool, error) {
 
 // Hash returns the root hash.
 func (t *ImmutableTree) Hash() ([]byte, error) {
-	hash, _, err := t.root.hashWithCount()
-	return hash, err
+	return t.root.hashWithCount(t.version + 1)
```
Comment (tac0turtle) [inline immutable_tree.go:152]: why + 1?

File: import_test.go [hunk 3]
```diff
@@ -160,7 +161,7 @@ func TestImporter_Close(t *testing.T) {
 	importer, err := tree.Import(1)
 	require.NoError(t, err)
 
-	err = importer.Add(&ExportNode{Key: []byte("key"), Value: []byte("value"), Version: 1, Height: 0})
+	err = importer.Add(&ExportNode{Key: []byte("key"), Value: []byte("value"), NodeKey: &NodeKey{version: 1, path: big.NewInt(1)}, Height: 0})
```
Comment (tac0turtle) [inline import_test.go:164]: how does this work for existing exports?or is the migration path planned to be different ?

File: mutable_tree.go [hunk 4]
```diff
@@ -74,34 +71,32 @@ func (tree *MutableTree) IsEmpty() bool {
 
 // VersionExists returns whether or not a version exists.
 func (tree *MutableTree) VersionExists(version int64) bool {
-	tree.mtx.Lock()
-	defer tree.mtx.Unlock()
-
-	if tree.allRootLoaded {
-		return tree.versions[version]
+	firstVersion, err := tree.ndb.getFirstVersion()
```
Comment (tac0turtle) [inline mutable_tree.go:74]: is there not a need to lock here anymore?

**Comment by cool-develope** [comment]
This PR has two problems, the `export/import` is not working with the original version.
It violates #660.

File: immutable_tree.go (hunk 2 above)
Comment (cool-develope) [reply immutable_tree.go:152]: This version is for non-persisted nodes, the current version is not updated yet.

File: import_test.go (hunk 3 above)
Comment (cool-develope) [reply import_test.go]: I believe it would look different, please refer #662

File: mutable_tree.go (hunk 4 above)
Comment (cool-develope) [reply mutable_tree.go]: yeah, we don't use the map of versions,since we are keeping the range of versions (fristVersion, latestVersion)

**Pull request closed without merging**